docs/chainlink-keepers/flexible-upkeeps.md

```diff
@@ -0,0 +1,157 @@
+
+1. Deploy the contract using Remix on the [supported testnet](../supported-networks) of your choice.
+
+1. Before registering the upkeep for your contract, decrease the balances of some elements. Still on Remix:
+   Withdraw 100 at 10,100,300,350,500,600,670,700,900. Pass the following to the withdraw function: 100,[10,100,300,350,500,600,670,700,900]
+
+   ![Withdraw 100 at 10,100,300,350,500,600,670,700,900](/images/contract-devs/keeper/balancerOnChain-withdraw.png)
+
+   Note that you can also perform this step after registering the upkeep.
+
+1. Register the upkeep for your contract as explained [here](/docs/chainlink-keepers/register-upkeep/). Because this example has high gas requirements, specify the maximum allowed gas limit of 2,500,00.
+
+1. Once registration is confirmed, you will notice that an upkeep has been done.
+
+   ![BalancerOnChain Upkeep History](/images/contract-devs/keeper/balancerOnChain-history.png)
+
+1. Click the transaction hash to see the transaction details in Etherscan. There you can find how much gas was used in the upkeep transaction.
+
+   ![BalancerOnChain Gas](/images/contract-devs/keeper/balancerOnChain-gas.png)
+
+In this example, the `performUpkeep()` function used **2,481,379** gas. This example has two main issues:
+
+- All computation is done in `performUpkeep()` , which is a state modifying function, leading to high gas consumption.
+- This example is simple but looping over large arrays with state updates can cause the transaction to hit the gas limit of the [network](../supported-networks) , preventing the performUpkeep from running successfully.  
+  In the next [section](#solution-perform-complex-computations-with-no-gas-fees), you will leverage the flexibility of Chainlink Keepers to reduce gas fees and mitigate the risks of running out of gas.
+
+## Solution: Perform complex computations with no gas fees
+
+In this section, we will slightly modify the contract and move the computation to `checkUpkeep()` function, which <ins>doesn’t consume any gas</ins>. Moreover, we will support multiple upkeeps for the same contract to parallelize the work to be done.
+The main differences with the previous contract are the following (cf. contract code below):
+
+- The `checkUpkeep()` function receives [checkData](/docs/chainlink-keepers/compatible-contracts/#checkdata), which is used to pass arbitrary bytes to the function. In this case, we will pass a `lowerBound` and an `upperBound` to scope the work to a subarray of `balances`. This will allow us to create several upkeeps with different values of `checkData`. The function loops over the subarray and looks for the indexes of the elements which require balancing and calculates the required `increments`. It then returns `upkeepNeeded == true`, and `performData` , which is calculated by encoding `indexes` and `increments`. Note that `checkUpkeep()` is a view function, meaning that any computation does not consume any gas.
```

Comment:
```suggestion
- The `checkUpkeep()` function receives [`checkData`](/docs/chainlink-keepers/compatible-contracts/#checkdata), which is used to pass arbitrary bytes to the function. In this case, we will pass a `lowerBound` and an `upperBound` to scope the work to a subarray of `balances`. This will allow us to create several upkeeps with different values of `checkData`. The function loops over the subarray and looks for the indexes of the elements which require balancing and calculates the required `increments`. It then returns `upkeepNeeded == true`, and `performData` , which is calculated by encoding `indexes` and `increments`. Note that `checkUpkeep()` is a view function, meaning that any computation does not consume any gas.
```
```suggestion
- The `checkUpkeep()` function receives [checkData](/docs/chainlink-keepers/compatible-contracts/#checkdata), which is used to pass arbitrary bytes to the function. In this case, we will pass a `lowerBound` and an `upperBound` to scope the work to a subarray of `balances`. This will allow us to create several upkeeps with different values of `checkData`. The function loops over the subarray and looks for the indexes of the elements which require balancing and calculates the required `increments`. It then returns `upkeepNeeded == true`, and `performData` , which is calculated by encoding `indexes` and `increments`. Note that `checkUpkeep()` is a view function, meaning that any computation does not consume any gas.
```